File: forjs/python/quick_sort.py

def partition(ar:list[int],low:int,high:int) -> int:
    pivot = ar[high]
    i = low - 1
    for j in range(low,high): 
        if ar[j]<pivot:
            i += 1 
            ar[i],ar[j]=ar[j],ar[i]
    ar[i+1],ar[high] = ar[high],ar[i+1]
    return i + 1

def quick_sort(ar: list[int], low=None,high=None)->None:
    low = 0 if low==None else low
    high = len(ar)-1 if high==None else high
    if(low<high):
        pi :int= partition(ar,low,high)
        quick_sort(ar,low,pi-1)
        quick_sort(ar,pi+1,high)
ar = [5,4,3,2,1]
quick_sort(ar)
print( ar )

File: forjs/JavaScript/quick_sort.js
const partition = (arr, low, high) => {
  let pivot = arr[high];
  let i = low - 1;
  for (let j = low; j < high; j++) {
    if (arr[j] < pivot) {
      i++;
      [arr[j], arr[i]] = [arr[i], arr[j]];
    }
  }
  [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
  return i + 1;
};

function quick_sort(arr, low, high) {
  if (!low) low = 0;
  if (!high) high = arr.length - 1;

  if (low < high) {
    let pi = partition(arr, low, high);
    quick_sort(arr, low, pi - 1);
    quick_sort(arr, pi + 1, high);
  }
}

function quickSort(array){
  quick_sort(array)
  return array
}
let array = [5,4,3,2,1]
console.log(quickSort(array));
